Migrate to zlib-ng, part 2: consume it in runtime (second attempt)

carlossanlop opened the pull request:
Contributes to: https://github.com/dotnet/runtime/issues/101465

This PR reapplies the reverted changes due to a wasm break in the official build: https://github.com/dotnet/runtime/pull/102403

It also includes an extra commit that fixes the NativeAOT build failure that only happened in the official build.

carlossanlop committed:
Reapply "Migrate to zlib-ng, part 2: consume it in runtime (#102403)" (#104414)

This reverts commit 5b86dca8e75627d1f5ad77f1dcffafa96e4f5049.

carlossanlop committed:
Unblock wasm build by excluding it from using zlib-ng and instead use the system installed zlib

carlossanlop commented on src/mono/CMakeLists.txt:
```diff
@@ -510,7 +510,7 @@ endif()
+  if (CLR_CMAKE_USE_SYSTEM_ZLIB OR CLR_CMAKE_HOST_ARCH_WASM)
     # if we're not on a platform where we use the in-tree zlib-ng, require system zlib
     include(${CLR_SRC_NATIVE_DIR}/libs/System.IO.Compression.Native/extra_libs.cmake)
     set(Z_LIBS)
```
Here is where the value of `Z_LIBS` is set for all the CMakeLists.txt in the subfolders to use.

carlossanlop replied:
Here is where the value of Z_LIBS is set for all the CMakeLists.txt in the subfolders to use.

jkotas commented on src/native/external/zlib-ng.cmake:
```diff
@@ -15,12 +15,14 @@ add_compile_options($<$<COMPILE_LANG_AND_ID:C,MSVC>:/wd4127>) # warning C4127: c
 add_compile_options($<$<COMPILE_LANG_AND_ID:C,MSVC>:/wd4242>) # 'function': conversion from 'unsigned int' to 'Pos', possible loss of data, in various deflate_*.c files
 add_compile_options($<$<COMPILE_LANG_AND_ID:C,MSVC>:/wd4244>) # 'function': conversion from 'unsigned int' to 'Pos', possible loss of data, in various deflate_*.c files
 
+# NOTE: Wasm is temporarily using system zlib, but will be switched to zlib-ng eventually. This branch should not affect the build.
```
I would either delete this comment so that this file does not need it to be touched when re-enabling; or delete this whole block and add it back as part re-enabling.

carlossanlop replied:
I added the comments in this file without commenting or removing the lines that affect wasm because we are now supposed to never import zlib-ng for wasm.

carlossanlop commented on src/native/libs/System.IO.Compression.Native/extra_libs.cmake:
```diff
@@ -1,8 +1,8 @@
-  if (CLR_CMAKE_TARGET_BROWSER OR HOST_BROWSER OR CLR_CMAKE_TARGET_WASI OR HOST_WASI)
-      # nothing special to link
+  if (CLR_CMAKE_HOST_ARCH_WASM OR CLR_CMAKE_TARGET_BROWSER OR HOST_BROWSER OR CLR_CMAKE_TARGET_WASI OR HOST_WASI)
+      list(APPEND ZLIB_LIBRARIES z)
```
I could not use `find_package( ZLIB REQUIRED)` before line 5 because I would get this error:

```
  CMake Error at /usr/share/cmake-3.28/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
    Could NOT find ZLIB (missing: ZLIB_LIBRARY ZLIB_INCLUDE_DIR)
  Call Stack (most recent call first):
    /usr/share/cmake-3.28/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
    /usr/share/cmake-3.28/Modules/FindZLIB.cmake:199 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
    /home/carlos/repos/runtime/src/native/libs/System.IO.Compression.Native/extra_libs.cmake:5 (find_package)
    CMakeLists.txt:517 (append_extra_compression_libs)
```

In my docker container that imitates what we do in the CI, I can confirm that libz is available under the crossrootfs folder:

```
# ls -l /crossrootfs/x64/lib/x86_64-linux-gnu/*libz*
lrwxrwxrwx 1 root root     13 Jan 21  2020 /crossrootfs/x64/lib/x86_64-linux-gnu/libz.so.1 -> libz.so.1.2.8
-rw-r--r-- 1 root root 104864 Jan 21  2020 /crossrootfs/x64/lib/x86_64-linux-gnu/libz.so.1.2.8
```

So I figured maybe this is a similar case to the android branch, where find_package would conflict, so I tried by just adding the `list` call too. Am I missing something?

carlossanlop commented on src/mono/mono/metadata/CMakeLists.txt:
```diff
@@ -205,6 +205,8 @@ if(HOST_WIN32 AND NOT DISABLE_SHARED_LIBS)
     if (NOT CLR_CMAKE_USE_SYSTEM_ZLIB)
         list(APPEND METADATAOBJECTSSHARED_LINKABLE_LIBS zlib)
+    elseif(CLR_CMAKE_HOST_ARCH_WASM)
+        list(APPEND METADATAOBJECTSSHARED_LINKABLE_LIBS ${Z_LIBS})
```
I added a `message(FATAL_ERROR "...")` here and confirmed this is where we attempt to link zlib to metadata_objects, right before the build fails with:

```
 [ 63%] Building C object mono/metadata/CMakeFiles/metadata_objects.dir/debug-mono-ppdb.c.o
  /home/carlos/repos/runtime/src/mono/mono/metadata/debug-mono-ppdb.c:33:10: fatal error: 'zlib.h' file not found
  #include <zlib.h>
           ^~~~~~~~
  1 error generated.
  emcc: error: '/home/carlos/repos/runtime/src/mono/browser/emsdk/bin/clang -target wasm32-unknown-emscripten -fignore-exceptions -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -DEMSCRIPTEN -Werror=implicit-function-declaration --sysroot=/home/carlos/repos/runtime/src/mono/browser/emsdk/emscripten/cache/sysroot -Xclang -iwithsysroot/include/fakesdl -Xclang -iwithsysroot/include/compat -DCOMPILER_SUPPORTS_W_RESERVED_IDENTIFIER -DHAVE_CONFIG_H -DHAVE_SGEN_GC -DMONO_DLL_EXPORT -D_THREAD_SAFE -I/home/carlos/repos/runtime/artifacts/obj -I/home/carlos/repos/runtime/src/native -I/home/carlos/repos/runtime/artifacts/obj/mono/browser.wasm.Release/mono/metadata/../.. -I/home/carlos/repos/runtime/src/mono/mono/metadata/../.. -I/home/carlos/repos/runtime/src/mono/mono/metadata/.. -I/home/carlos/repos/runtime/src/native/public/. -I/home/carlos/repos/runtime/artifacts/obj/mono/browser.wasm.Release/mono/eglib -I/home/carlos/repos/runtime/src/mono/mono/eglib -fno-strict-aliasing -fwrapv -Wall -Wunused -Wmissing-declarations -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Wno-format-zero-length -Wno-unused-function -Qunused-arguments -Wno-tautological-compare -Wno-parentheses-equality -Wno-self-assign -Wno-return-stack-address -Wno-constant-logical-operand -Wno-zero-length-array -Wno-asm-operand-widths -Wmissing-prototypes -Wstrict-prototypes -Wnested-externs -Werror=return-type -Werror=implicit-function-declaration -Werror=incompatible-pointer-types -O3 -DNDEBUG -std=gnu11 -g3 -fPIC -fvisibility=hidden -Wno-strict-prototypes -Wno-unused-but-set-variable -Wno-single-bit-bitfield-constant-conversion -Os -ffp-contract=off -MD -MT mono/metadata/CMakeFiles/metadata_objects.dir/debug-mono-ppdb.c.o -MF CMakeFiles/metadata_objects.dir/debug-mono-ppdb.c.o.d -c /home/carlos/repos/runtime/src/mono/mono/metadata/debug-mono-ppdb.c -o CMakeFiles/metadata_objects.dir/debug-mono-ppdb.c.o' failed (returned 1)
  make[2]: *** [mono/metadata/CMakeFiles/metadata_objects.dir/build.make:257: mono/metadata/CMakeFiles/metadata_objects.dir/debug-mono-ppdb.c.o] Error 1
  make[1]: *** [CMakeFiles/Makefile2:415: mono/metadata/CMakeFiles/metadata_objects.dir/all] Error 2
  make: *** [Makefile:136: all] Error 2
/home/carlos/repos/runtime/src/mono/mono.proj(765,5): error MSB3073: The command "cmake --build . --target install --config Release --parallel 12" exited with code 2.

Build FAILED.
```

carlossanlop commented on src/native/external/zlib-ng.cmake:
```diff
@@ -15,12 +15,14 @@ add_compile_options($<$<COMPILE_LANG_AND_ID:C,MSVC>:/wd4127>) # warning C4127: c
 endif()
 
-set(BUILD_SHARED_LIBS OFF) # Shared libraries aren't supported in wasm
+set(BUILD_SHARED_LIBS OFF) # Shared libraries aren't supported in wasm, and even though it consumes system zlib, it will be switched to zlib-ng eventually.
```
Since we plan on adding zlib-ng support to wasm _eventually_ (this is just temporary), I decided to not remove this line that disables shared libs, otherwise I would have to make a ton of changes to the PR.

carlossanlop reviewed (commented):
@pavelsavara @ilonatommy @kg @jkotas @jkoritzinsky @akoeplinger  I need some help, I am still unable to get a successful  build in wasm.

I can repro the exact same scenario used in the CI by executing these commands inside my Ubuntu WSL:

```
docker run -ti -d -v /home/carlos/repos/runtime:/home/carlos/repos/runtime -e ROOTFS_DIR=/crossrootfs/x64 -w /home/carlos/repos/runtime  mcr.microsoft.com/dotnet-buildtools/prereqs:azurelinux-3.0-webassembly-amd64-net9.0
docker attach <GUID_HERE>

git config --global --add safe.directory /home/carlos/repos/runtime
./build.sh mono+libs+host+packs+libs.tests -c Release -arch wasm -os browser /p:MonoEnableAssertMessages=true /p:BrowserHost=linux /p:AotHostArchitecture=x64 /p:AotHostOS=linux
```

And the build is failing with the following error, which tells me that I am still unable to successfully link the system zlib (which I confirmed exists inside the crossrootfs folder) to mono/metadata:

```
 /home/carlos/repos/runtime/src/mono/mono/metadata/debug-mono-ppdb.c:33:10: fatal error: 'zlib.h' file not found
  #include <zlib.h>
           ^~~~~~~~
  1 error generated.
  emcc: error: '/home/carlos/repos/runtime/src/mono/browser/emsdk/bin/clang -target wasm32-unknown-emscripten -fignore-exceptions -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -DEMSCRIPTEN -Werror=implicit-function-declaration --sysroot=/home/carlos/repos/runtime/src/mono/browser/emsdk/emscripten/cache/sysroot -Xclang -iwithsysroot/include/fakesdl -Xclang -iwithsysroot/include/compat -DCOMPILER_SUPPORTS_W_RESERVED_IDENTIFIER -DHAVE_CONFIG_H -DHAVE_SGEN_GC -DMONO_DLL_EXPORT -D_THREAD_SAFE -I/home/carlos/repos/runtime/artifacts/obj -I/home/carlos/repos/runtime/src/native -I/home/carlos/repos/runtime/artifacts/obj/mono/browser.wasm.Release/mono/metadata/../.. -I/home/carlos/repos/runtime/src/mono/mono/metadata/../.. -I/home/carlos/repos/runtime/src/mono/mono/metadata/.. -I/home/carlos/repos/runtime/src/native/public/. -I/home/carlos/repos/runtime/artifacts/obj/mono/browser.wasm.Release/mono/eglib -I/home/carlos/repos/runtime/src/mono/mono/eglib -fno-strict-aliasing -fwrapv -Wall -Wunused -Wmissing-declarations -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Wno-format-zero-length -Wno-unused-function -Qunused-arguments -Wno-tautological-compare -Wno-parentheses-equality -Wno-self-assign -Wno-return-stack-address -Wno-constant-logical-operand -Wno-zero-length-array -Wno-asm-operand-widths -Wmissing-prototypes -Wstrict-prototypes -Wnested-externs -Werror=return-type -Werror=implicit-function-declaration -Werror=incompatible-pointer-types -O3 -DNDEBUG -std=gnu11 -g3 -fPIC -fvisibility=hidden -Wno-strict-prototypes -Wno-unused-but-set-variable -Wno-single-bit-bitfield-constant-conversion -Os -ffp-contract=off -MD -MT mono/metadata/CMakeFiles/metadata_objects.dir/debug-mono-ppdb.c.o -MF CMakeFiles/metadata_objects.dir/debug-mono-ppdb.c.o.d -c /home/carlos/repos/runtime/src/mono/mono/metadata/debug-mono-ppdb.c -o CMakeFiles/metadata_objects.dir/debug-mono-ppdb.c.o' failed (returned 1)
  make[2]: *** [mono/metadata/CMakeFiles/metadata_objects.dir/build.make:257: mono/metadata/CMakeFiles/metadata_objects.dir/debug-mono-ppdb.c.o] Error 1
  make[1]: *** [CMakeFiles/Makefile2:415: mono/metadata/CMakeFiles/metadata_objects.dir/all] Error 2
  make: *** [Makefile:136: all] Error 2
/home/carlos/repos/runtime/src/mono/mono.proj(765,5): error MSB3073: The command "cmake --build . --target install --config Release --parallel 12" exited with code 2.

Build FAILED.
```

jkotas commented on eng/native/configureplatform.cmake:
```diff
@@ -500,3 +500,9 @@ if(LOWERCASE_CMAKE_BUILD_TYPE STREQUAL debug)
 endif()
+
+if (CLR_CMAKE_TARGET_ANDROID OR CLR_CMAKE_TARGET_MACCATALYST OR CLR_CMAKE_TARGET_IOS OR CLR_CMAKE_TARGET_TVOS OR CLR_CMAKE_HOST_ARCH_WASM OR CLR_CMAKE_HOST_ARCH_ARMV6)
+    # Our zlib-ng copy should only be used in non-mobile and non-browser platforms only.
+    # In mobile platforms we should use the zlib package provided by the system.
```
```suggestion
    # Some platforms are opted-out from using the in-tree zlib-ng by default:
    # Android and iOS-like platforms: concerns about extra binary size
    # Wasm: zlib-ng breaks the build
    # Armv6: zlib-ng has build breaks
```
This comment is not 100% correct. Wasm is not the same as browser, and Armv6 is not really a mobile platform. I do not think that comments that try to rephrase the code are useful. Good comments explain why it is written like that.

carlossanlop replied:
Makes sense, thanks. I'll apply this suggestion when I get a proper fix for the wasm failure to avoid resetting the CI unnecessarily.

carlossanlop commented:
@LoopedBard3 I think I need to remove `zlib` from the list of dependencies that get installed by this performance job. Can you please confirm? If yes, what azp run do I need to execute to confirm that change?:

https://github.com/dotnet/runtime/tree/main/eng/pipelines/coreclr/templates/run-performance-job.yml#L114

carlossanlop committed:
Fix windows nativeaot failure happening when executing:

build.cmd -ci -arch x64 -os windows  -s clr.nativeaotlibs+clr.nativeaotruntime+libs+packs -c Release /p:BuildNativeAOTRuntimePack=true /p:SkipLibrariesNativeRuntimePackages=true

Failure message this commit fixes:

[.packages\microsoft.dotnet.sharedframework.sdk\9.0.0-beta.24327.1\targets\sharedfx.targets(296,5):
error : The following files are missing entries in the templated manifest:
error : zlibstatic.lib. Add these file names with extensions to the 'PlatformManifestFileEntry' item group for the runtime pack and corresponding ref pack to include them in the platform manifest.
[\src\installer\pkg\sfx\Microsoft.NETCore.App\Microsoft.NETCore.App.Runtime.sfxproj]

carlossanlop committed:
Apply jkotas comment suggestion in configureplatform.cmake

carlossanlop committed:
Delete unnecessary comment in zlib-ng.cmake

caaavik-msft commented:
> @LoopedBard3 I think I need to remove `zlib` from the list of dependencies that get installed by this performance job. Can you please confirm? If yes, what azp run do I need to execute to confirm that change?:
> 
> https://github.com/dotnet/runtime/blob/main/eng/pipelines/coreclr/templates/run-performance-job.yml#L114

It would be "dotnet-runtime-perf", and this change would only affect the linux_musl jobs. Currently that job is failing though intermittently due to another issue we are still investigating but I can confirm the changes if you run the pipeline with that fix.

carlossanlop commented on src/mono/mono/eventpipe/test/CMakeLists.txt:
```diff
@@ -42,7 +42,7 @@ if(ENABLE_PERFTRACING)
         set(EPTEST_LINKABLE_LIBS "")
         list(APPEND EPTEST_LINKABLE_LIBS eglib_api monosgen-static ${OS_LIBS} ${LLVM_LIBS} monoapi)
-        if (CLR_CMAKE_USE_SYSTEM_ZLIB)
+        if (CLR_CMAKE_USE_SYSTEM_ZLIB OR CLR_CMAKE_HOST_ARCH_WASM)
```
This extra condition is unnecessary here. CLR_CMAKE_USE_SYSTEM_ZLIB should include wasm.

carlossanlop commented:
I am having trouble getting wasm to find the installed zlib. I cannot use find_package in extra_libs.cmake, as it does not know where libz.a is found. I know it can be available inside the CROSS_ROOTFS if the CI leg creates the docker image with the necessary argument to enable such folder. But I am noticing that many wasm/wasi CI legs do _not_ set a CROSS_ROOTFS directory, so that's not an option.

The main reason why this change was reverted was because of a NativeAOT failure in official builds ([proof](https://github.com/dotnet/runtime/pull/104414#issuecomment-2210988895)), _not_ a wasm failure.

The wasm problem was mentioned in the revert PR as an additional reason to support the revert, but it was only a local build problem, not an official build problem, and it seemed the root cause was not related to my PR, but to additional changes that were being tested locally.

I already fixed the NativeAOT problem with commit https://github.com/dotnet/runtime/pull/104454/commits/d232aaf5e3305ecff13e22fa0a52fece58850a6f, which would unblock the official build failure. But I'm now concerned that we are adding many more workarounds and exclusions for wasm/wasi that seem to be even more problematic than just enabling zlib-ng, which was already working fine. We need the zlib-ng change merged ASAP to ensure we see the zlib-ng behavior not only in runtime but in dependent repos, before we snap Preview7. This revert is severely preventing us from achieving that.

@pavelsavara @ilonatommy @kg I am leaning towards reverting commit https://github.com/dotnet/runtime/pull/104454/commits/2c60469cd737ec86770158e752181b0c5ba43b0d from this PR (which was attempting to exclude wasm/wasi from including zlib-ng), merging this PR with _just_ the nativeAOT fix for the official build, and then working with you all to determine the best course of action to help you get your local build to pass with your changes on top of this (because the official builds did not hit any wasm/wasi failures related to the zlib-ng change).

Thoughts?

carlossanlop committed:
Revert the wasm changes from 2c60469cd737ec86770158e752181b0c5ba43b0d because the local browser build is passing now.

carlossanlop commented:
/azp run list

carlossanlop commented:
/azp list

carlossanlop commented:
/azp run runtime-wasm

carlossanlop commented:
/azp run runtime-nativeaot-outerloop

carlossanlop commented:
/azp run runtime-community

carlossanlop commented:
/azp run runtime-extra-platforms

pavelsavara commented:
`zlib-ng` is not ready for WASI Preview2, see https://github.com/dotnet/runtime/pull/104414#issuecomment-2214815763

ericstj commented:
> `zlib-ng` is not ready for WASI Preview2, see [#104414 (comment)](https://github.com/dotnet/runtime/pull/104414#issuecomment-2214815763)

Let's collaborate on getting those fixed in the WASI PR.  It will be easier to test and iterate there.  It's important that this change makes it into Preview 7.

ericstj commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props:
```diff
@@ -279,6 +279,7 @@
     <PlatformManifestFileEntry Include="libicuuc.a" IsNative="true" />
     <!-- zlib-specific files -->
     <PlatformManifestFileEntry Include="libz.a" IsNative="true" />
+    <PlatformManifestFileEntry Include="zlibstatic.lib" IsNative="true" />
```
is it intentional to *ship* zlibstatic.lib in our shared framework?  That seems unusual as it's a dev-time resource - however I see other dev-time resources here like `main.c`, `driver.h`, etc.  I take it you needed to add this to build - but I bet a better fix is to get all of this stuff out of the platform manifest since I don't think it needs to be fed to the SDK for conflict resolution @jkoritzinsky @dsplaisted.

I see we have a metadata flag `ExcludeFromDataFiles` for that: https://github.com/dotnet/arcade/blob/76f733ee57811c38bb5b8e1ac9c6c50e92bc5dc9/src/Microsoft.DotNet.SharedFramework.Sdk/targets/sharedfx.targets#L249

I think we should review all the things added here, remove files that don't need to ship - if any, and apply `ExcludeFromDataFiles` for the files that need to ship and don't actually need to go into the manifest.  That can be done separately.

carlossanlop replied:
I'll keep this in mind as a follow up.

jkotas replied:
> is it intentional to ship zlibstatic.lib in our shared framework?

zlibstatic.lib ships in the native aot runtime package only. It does not ship in the regular shared framework.

jkoritzinsky replied:
When I implemented the shared framework SDK, no one really knew what needed to go into the manifest, so it requires/lists all files. If we know what types of files need to be in it, I can update the SDK to only validate (for the template case) or include (for the generative case) files that matter.

ericstj reviewed (approved):
Approve this change to get this in and building.  Please file a follow up issue on the platform manifest cleanup, and work with @pavelsavara on wasi fix.

carlossanlop commented:
/ba-g All failures determined to be unrelated.

carlossanlop commented:
@LoopedBard3 @caaavik-msft @DrewScoggins where can I see the runtime perf chart before and after this change?

LoopedBard3 commented:
If you have specific tests in mind, we have an ADX dashboard and database. I have DMed these to you. Otherwise, if you want to just look at a few of the test results, we also have an [allTestHistory](https://pvscmdupload.z22.web.core.windows.net/reports/allTestHistory/TestHistoryIndexIndex.html) that is updated once a day with the latest run data. Although it seems we currently have a pathing bug so you will need to remove the "reports/" from the url to view the actual test if the links give a 404.

am11 commented:
@carlossanlop, VMR jobs are failing with:

```
      Generating native code
      /usr/bin/ld.bfd: /vmr/src/runtime/artifacts/bin/microsoft.netcore.app.runtime.linux-x64/Release/runtimes/linux-x64/native/libSystem.IO.Compression.Native.a(pal_zlib.c.o): in function `CompressionNative_DeflateInit2_':
      /vmr/src/runtime/src/native/libs/System.IO.Compression.Native/pal_zlib.c:122:(.text.CompressionNative_DeflateInit2_+0x89): undefined reference to `deflateInit2_'
      /usr/bin/ld.bfd: /vmr/src/runtime/artifacts/bin/microsoft.netcore.app.runtime.linux-x64/Release/runtimes/linux-x64/native/libSystem.IO.Compression.Native.a(pal_zlib.c.o): in function `CompressionNative_Deflate':
      /vmr/src/runtime/src/native/libs/System.IO.Compression.Native/pal_zlib.c:134:(.text.CompressionNative_Deflate+0x2c): undefined reference to `deflate'
      /usr/bin/ld.bfd: /vmr/src/runtime/artifacts/bin/microsoft.netcore.app.runtime.linux-x64/Release/runtimes/linux-x64/native/libSystem.IO.Compression.Native.a(pal_zlib.c.o): in function `CompressionNative_DeflateEnd':
      /vmr/src/runtime/src/native/libs/System.IO.Compression.Native/pal_zlib.c:145:(.text.CompressionNative_DeflateEnd+0x29): undefined reference to `deflateEnd'
      /usr/bin/ld.bfd: /vmr/src/runtime/artifacts/bin/microsoft.netcore.app.runtime.linux-x64/Release/runtimes/linux-x64/native/libSystem.IO.Compression.Native.a(pal_zlib.c.o): in function `CompressionNative_InflateInit2_':
      /vmr/src/runtime/src/native/libs/System.IO.Compression.Native/pal_zlib.c:159:(.text.CompressionNative_InflateInit2_+0x6b): undefined reference to `inflateInit2_'
      /usr/bin/ld.bfd: /vmr/src/runtime/artifacts/bin/microsoft.netcore.app.runtime.linux-x64/Release/runtimes/linux-x64/native/libSystem.IO.Compression.Native.a(pal_zlib.c.o): in function `CompressionNative_Inflate':
      /vmr/src/runtime/src/native/libs/System.IO.Compression.Native/pal_zlib.c:171:(.text.CompressionNative_Inflate+0x2c): undefined reference to `inflate'
      /usr/bin/ld.bfd: /vmr/src/runtime/artifacts/bin/microsoft.netcore.app.runtime.linux-x64/Release/runtimes/linux-x64/native/libSystem.IO.Compression.Native.a(pal_zlib.c.o): in function `CompressionNative_InflateEnd':
      /vmr/src/runtime/src/native/libs/System.IO.Compression.Native/pal_zlib.c:182:(.text.CompressionNative_InflateEnd+0x29): undefined reference to `inflateEnd'
      /usr/bin/ld.bfd: /vmr/src/runtime/artifacts/bin/microsoft.netcore.app.runtime.linux-x64/Release/runtimes/linux-x64/native/libSystem.IO.Compression.Native.a(pal_zlib.c.o): in function `CompressionNative_Crc32':
      /vmr/src/runtime/src/native/libs/System.IO.Compression.Native/pal_zlib.c:192:(.text.CompressionNative_Crc32+0x8): undefined reference to `crc32'
    clang : error : linker command failed with exit code 1 (use -v to see invocation) [/vmr/src/runtime/src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj]
    /vmr/src/runtime/artifacts/bin/coreclr/linux.x64.Release/build/Microsoft.NETCore.Native.targets(368,5): error MSB3073: The command ""/usr/bin/clang-14" "/vmr/src/runtime/artifacts/obj/coreclr/crossgen2_publish/linux.x64.Release/native/crossgen2.o" -o "/vmr/src/runtime/artifacts/bin/crossgen2_publish/x64/Release/native/crossgen2" -Wl,--version-script=/vmr/src/runtime/artifacts/obj/coreclr/crossgen2_publish/linux.x64.Release/native/crossgen2.exports -Wl,--export-dynamic -gz=zlib -fuse-ld=bfd /vmr/src/runtime/artifacts/bin/coreclr/linux.x64.Release/aotsdk/libbootstrapper.o /vmr/src/runtime/artifacts/bin/coreclr/linux.x64.Release/aotsdk/libRuntime.ServerGC.a /vmr/src/runtime/artifacts/bin/coreclr/linux.x64.Release/aotsdk/libeventpipe-enabled.a /vmr/src/runtime/artifacts/bin/coreclr/linux.x64.Release/aotsdk/libRuntime.VxsortEnabled.a /vmr/src/runtime/artifacts/bin/coreclr/linux.x64.Release/aotsdk/libstandalonegc-disabled.a /vmr/src/runtime/artifacts/bin/coreclr/linux.x64.Release/aotsdk/libstdc++compat.a /vmr/src/runtime/artifacts/bin/coreclr/linux.x64.Release/aotsdk/libz.a /vmr/src/runtime/artifacts/bin/microsoft.netcore.app.runtime.linux-x64/Release/runtimes/linux-x64/native/libSystem.Native.a /vmr/src/runtime/artifacts/bin/microsoft.netcore.app.runtime.linux-x64/Release/runtimes/linux-x64/native/libSystem.IO.Compression.Native.a /vmr/src/runtime/artifacts/bin/microsoft.netcore.app.runtime.linux-x64/Release/runtimes/linux-x64/native/libSystem.Net.Security.Native.a /vmr/src/runtime/artifacts/bin/microsoft.netcore.app.runtime.linux-x64/Release/runtimes/linux-x64/native/libSystem.Security.Cryptography.Native.OpenSsl.a -g -Wl,-rpath,'$ORIGIN' -Wl,--build-id=sha1 -Wl,--as-needed -pthread -ldl -lrt -lm -pie -Wl,-pie -Wl,-z,relro -Wl,-z,now -Wl,--eh-frame-hdr -Wl,--discard-all -Wl,--gc-sections" exited with code 1. [/vmr/src/runtime/src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj]
```
blocking SDK https://github.com/dotnet/sdk/pull/42019. Do we need to set `-p:UseSystemZlib=true`?

carlossanlop commented:
Yep, I've been notified and I'm looking into this.

> Do we need to set -p:UseSystemZlib=true?

No, that property needs to be set conditionally.

LoopedBard3 commented:
Likely related improvements:
Linux arm64: https://github.com/dotnet/perf-autofiling-issues/issues/38054

stephentoub commented:
> We are seeing different results in the compressed payload in aspnetcore tests, 1 byte difference, is that expected with this switch? Wondering if a gzipped payload is supposed to be deterministic with the same compression level.

There's no guarantee about compressed bytes being exactly the same from version to version. It's expected that updating the zlib implementation may change the exact compressed bytes generated.
